### ISSUE TITLE
What all files in the public folder needs to be persisted for Incremental Build

### ISSUE THREAD

**Issue by sharanrprasad** [gatsbyjs/gatsby#22907]
Hello everyone. I was trying out the new Gatsby incremental build on one of our existing CI/CD pipelines. When ever a page is removed from our CMS incremental build fails. Should all the contents of the public folder be there for the build to pass.
Our site is hosted on S3. Currently I am persisting the cache between each builds using another S3 bucket. If all the contents of pubic folder from previous build also needs to be present then I have to pull content for about 10000 pages and that adds about 2-3 minutes to the build time. Is there a way that I can only pull some of the existing public folder contents or nothing at all but the cache?

Since we have multiple websites building from the same repo currently using Gatsby cloud is not an option.

**Comment by pieh** [comment]
This seems related to https://github.com/gatsbyjs/gatsby/issues/22641, assuming you mean the `GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES` experimental mode.

You could potentially restore just parts of `public` directory (if restoring it is causing so much overhead). In particular you for sure will need to restore `public/static/d` - that's just initial thing that I can think of that NEED to be there. But there can be more dragons, that's why we conservatively mention to preserve full `.cache` and `public` directories (4th bullet point in https://www.gatsbyjs.org/docs/page-build-optimizations-for-incremental-data-changes/#more-information). Depending on the setup (used plugins, any custom code) restoring just `public/static/d` (result of static queries) might not be enough - we just don't have enough data to be confident about which parts exactly of `public` need to be there in subsequent builds

What kind of build errors are you seeing? Is it something similar to
```
Can't resolve '../../public/static/d/<some_hash>.json' in '/path/to/src/components
```
?

**Comment by sharanrprasad** [comment]
Thank you for the reply. I get errors like this `no such file or directory, scandir '/var/lib/buildkite-agent/builds/buildkite-crimson-website-i-00d1ac99c7281314e-1/crimson-education/website-v4-incremental-build/public/mx/our-services/medview/mmi/faqs'`

I have already persisted static folder between builds and that solved errors I was having. This error always comes up when a page is deleted. 

So the only solution is to have all the contents of the public folder between builds?

**Comment by StuartRayson** [comment]
@sharanrprasad  We had a similar issues storing and moving our built assets on S3 buckets, so we moved to use EFS and mounting

The public folder needs to be persisted as removing the  static folder can cause issues as @pieh mentioned. Currently we don't check if the files exists before removing as we assume the public folder is in the same state as the previous cache build. 

@pieh is this something we could add? 
I think only these functions need updating

https://github.com/gatsbyjs/gatsby/blob/a20d2556d11d9dc6774698154bc193990a06f9f9/packages/gatsby/src/commands/build-utils.ts#L49-L65

https://github.com/gatsbyjs/gatsby/blob/a20d2556d11d9dc6774698154bc193990a06f9f9/packages/gatsby/src/commands/build-utils.ts#L74-L92

**Comment by pieh** [comment]
> @pieh could we add these checks before attempting to remove the file/folder?
Ah yeah, we could do that for sure

The difficult part will be documenting this - as soon as start selectively restore `public` directory - the deployement to CDN will also need to be pretty custom - so you wouldn't be able to use something like `netlify deploy <public_dir>` (I don't remember exact command - just the commands that takes care of adding/removing/updating assets on CDN automatically will be out of the question) and instead would need to reach to lower level APIs to do this selectively.

So I think curse of action can be adding sanity `fs.exists(sync)` checks before deleting files/directories (as suggested by @StuartRayson ), but keep the documentation as-is (recommending full `public` restoration), because there's lot of dragons ahead when it comes to perf tuning around selective `public` dir restoration and CDN/asset buckets deployment

**Comment by sharanrprasad** [comment]
@StuartRayson Thank you for the reply. I will try with EFS. The above mentioned fix would be really great for huge sites like ours.

**Comment by pieh** [comment]
@StuartRayson's fix was published in `gatsby@2.20.28`
